src/python/WMCore/MicroService/Unified/RSEQuotas.py

```diff
@@ -0,0 +1,204 @@
+"""
+Non thread-safe object which provides all the RSE/PNN information
+required for automatic data placement.
+It can also communicate with other data management tools, like
+Detox, Rucio and PhEDEx.
+"""
+from __future__ import division, print_function
+
+from WMCore.MicroService.Unified.Common import getDetoxQuota
+from WMCore.MicroService.Unified.Common import getMSLogger
+
+
+class RSEQuotas(object):
```

Comment:
If I understood logic correctly you update info about specific rse and keep internal dict of them.
Why not to make a class which just keep info about single RSE instead. Then you can take whole logic of nodeUsage out of this class and use it in upper codebase making RSE updates concurrent/thread-safe if you want. The main problem is self.nodeUsage here, if you take it out and keep in another class then all RSE instances can be updated concurrently.

Reply:
That's the correct understand of this module, Valentin.
I'm not sure I followed your suggestion, especially regarding rse information lookup and update. In the end we would have to have a set of such objects (one for each RSE), but then we would have to scan every object in order to find the RSE that we want to (and then the update can be made).

Another possibility, as I quickly mentioned in the meeting, would be to add a locking system around the local cache updates. I'd need to educate myself on the best way to make it thread safe (maybe using the threading.Queue feature...). Still, that might have to wait for another iteration on the near future.